daemon/cmd/daemon_main.go

```diff
@@ -1766,14 +1766,17 @@ func runDaemon() {
 		log.WithError(err).Warn("Failed to send agent start monitor message")
 	}
 
-	// clean up all arp PERM entries that might have previously set by
-	// a Cilium instance
 	if !d.datapath.Node().NodeNeighDiscoveryEnabled() {
-		d.datapath.Node().NodeCleanNeighbors()
-	}
-	// Start periodical arping to refresh neighbor table
-	if d.datapath.Node().NodeNeighDiscoveryEnabled() && option.Config.ARPPingRefreshPeriod != 0 {
-		d.nodeDiscovery.Manager.StartNeighborRefresh(d.datapath.Node())
+		// Remove all non-GC'ed neighbor entries that might have previously set
+		// by a Cilium instance.
+		d.datapath.Node().NodeCleanNeighbors(false)
+	} else {
+		// If we came from an agent upgrade, migrate entries.
+		d.datapath.Node().NodeCleanNeighbors(true)
+		// Start periodical refresh of the neighbor table from the agent if needed.
+		if option.Config.ARPPingRefreshPeriod != 0 && !option.Config.ARPPingKernelManaged {
```

Comment:
nit... is there a reason to have the '&&' check seems ARPPingRefreshPeriod should be sufficient.

Is there a reason to have ARPPingRefreshPeriod != 0 and ARPPingKernelManaged set?

Reply:
ARPPingRefreshPeriod is set to 30s by default, unless the user explicitly turns it off. ARPPingKernelManaged is auto-detected inside the agent. So if available due to the probe succeeding, then we don't need to start the refresh controller at all.